test/metrics/prom-metrics-linter/metric_name_linter.go

```diff
@@ -62,21 +71,56 @@ func main() {
 		os.Exit(1)
 	}
 
+	// Collect and print metric problems first (sorted by metric name)
+	metricProblems := problems
 	for _, family := range families {
-		problems = CustomLinterRules(problems, family, operatorName, subOperatorName)
+		metricProblems = CustomMetricsValidation(metricProblems, family, operatorName, subOperatorName)
+	}
+	sort.Slice(metricProblems, func(i, j int) bool { return metricProblems[i].Metric < metricProblems[j].Metric })
+	for _, p := range metricProblems {
+		fmt.Printf("%s: %s\n", p.Metric, p.Text)
 	}
 
-	for _, problem := range problems {
-		fmt.Printf("%s: %s\n", problem.Metric, problem.Text)
+	// Then collect and print recording rule problems (sorted by rule name), skipping allowlisted names
+	allow := map[string]struct{}{}
+	if len(embeddedAllowlist) > 0 {
+		var cfg allowlistConfig
+		if err := json.Unmarshal(embeddedAllowlist, &cfg); err != nil {
+			panic("failed to read the allow list; " + err.Error())
+		}
+		for _, n := range cfg.Operators[subOperatorName] {
+			allow[n] = struct{}{}
+		}
+	}
+	ruleProblems := []promlint.Problem{}
+	for _, rr := range recordingRules {
+		if _, ok := allow[rr.Record]; !ok {
+			ruleProblems = append(ruleProblems, CustomRecordingRuleValidation(rr)...)
+		}
+	}
+	sort.Slice(ruleProblems, func(i, j int) bool { return ruleProblems[i].Metric < ruleProblems[j].Metric })
+	for _, p := range ruleProblems {
+		fmt.Printf("%s: %s\n", p.Metric, p.Text)
 	}
 }
 
-func parseMetricFamilies(jsonStr string) []*dto.MetricFamily {
-	var families []*dto.MetricFamily
-	err := json.Unmarshal([]byte(jsonStr), &families)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "error parsing metric families: %v\n", err)
+type recordingRule struct {
+	Record string `json:"record"`
+	Expr   string `json:"expr"`
+	Type   string `json:"type,omitempty"`
+}
+
+type inputJSON struct {
+	MetricFamilies []*dto.MetricFamily `json:"metricFamilies"`
+	RecordingRules []recordingRule     `json:"recordingRules"`
+}
+
+func parseInput(jsonStr string) ([]*dto.MetricFamily, []recordingRule) {
```

Comment:
Please only exit from main. Return error instead of exit. Then check the error in main and exit from there.